### PR TITLE
[ES-49] Add multiple signer support

### PR DESCRIPTION
**Pull request by bgravenorst**
Signed-off-by: Byron Gravenorst byron.gravenorst@consensys.net

## Pull Request Description
Add support for multiple signers.

## Fixed Issue(s)
ES-49

### CODE REVIEW

File: docs/HowTo/Transactions/Make-Transactions.md [hunk 1]
```diff
@@ -33,7 +33,7 @@ Transaction object for private transactions:
 
 | Key             | Type                | Required/Optional                  | Value                                                                                                                         |
 |-----------------|--:-:----------------|----------------------------------- |-------------------------------------------------------------------------------------------------------------------------------|
-| **from**        | Data, 20&nbsp;bytes | Required                           | Address of the sender. Must be the address of the keystore account.                                                                                                        |
+| **from**        | Data, 20&nbsp;bytes | Required                           | Address of the sender. Must be the address of the keystore account. If using [multiple key files](../../Concepts/Multiple-Key-Files.md), the keystore file name must contain the sender's address.                           |
```
Comment (lucassaldanha) [inline docs/HowTo/Transactions/Make-Transactions.md]: I think it isn't necessary to link the field definitions to the multi-file rules. I wouldn't mention it here.

File: docs/HowTo/Transactions/Make-Transactions.md [hunk 2]
```diff
@@ -105,7 +105,7 @@ Transaction object:
 
 | Key          | Type                | Required/Optional              | Value                                                                                  |
 |--------------|-:-:-----------------|--------------------------------|----------------------------------------------------------------------------------------|
-| **from**     | Data, 20&nbsp;bytes | Required                       | Address of the sender.                                                                 |
+| **from**     | Data, 20&nbsp;bytes | Required                       | Address of the sender. If using [multiple key files](../../Concepts/Multiple-Key-Files.md), the keystore file name must contain the sender's address.                                                                |
```
Comment (lucassaldanha) [inline docs/HowTo/Transactions/Make-Transactions.md]: Same as above. We don't need to mention this here.

**Comment by lucassaldanha** [comment]
> I don't know why you guys have a .DS_Store (docs/.DS_Store) file in your repo, it should probably be removed and added to the .gitignore file.

I've opened a PR to fix it #27

File: docs/Concepts/Multiple-Key-Files.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+# Store Multiple Key Files
+
+EthSigner supports file-based signing using multiple V3 keystore files. Each account that signs transactions on the network requires a key file. 
+
+The key file must follow the appropriate naming convention, and have an accompanying password file. [Create the required files in a directory](../Tutorials/Multifile.md) and use the [multifile-based-signer](../Reference/CLI/CLI-Syntax.md#multifile-options) sub-command to specify the directory location. 
+
+!!! note
+    EthSigner does not support using multiple key signers with Hashicorp Vault or Azure Key Vault.
+
+The key file name must contain the public key address and follow the `[<prefix>]<accountAddress>.key` format. The password file must have an identical name and end with the `.password` extension. No two key files can have the same key address in the file name, even if the prefix differs.
```
Comment (MadelineMurray) [inline docs/Concepts/Multiple-Key-Files.md]: The term `public key address` is a bit confusing - is it the public key (what you get from the subcommand `public-key export`) or is it the account address (what you get from the subcommand `public-key export-address`)?  

Ironically, the `export-address` description also uses the term `public key address` so it was probably me that introduced this to start with. As part of the rework, could you update it there (https://besu.hyperledger.org/en/stable/Reference/CLI/CLI-Subcommands/#export-address) as well.  ie, change `public key address` to `account address`.

File: docs/Concepts/Multiple-Key-Files.md [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+In the example the account address is prefixed with the date.
+
+!!! example
+    `10-10-2019_78e6e236592597c09d5c137c2af40aecd42d12a2.key`
+    
+    `10-10-2019_78e6e236592597c09d5c137c2af40aecd42d12a2.password`
+    
+!!! tip
+    Use the [`export-address`](https://besu.hyperledger.org/en/latest/Reference/CLI/CLI-Subcommands/#export-address) Hyperledger Besu command to obtain the public key address of the node.
```
Comment (MadelineMurray) [inline docs/Concepts/Multiple-Key-Files.md]: Hyperledger Besu command -> Hyperleder Besu subcommand

File: docs/Reference/CLI/CLI-Syntax.md [hunk 5]
```diff
@@ -282,4 +284,19 @@ Name of the vault to access. Sub-domain of `vault.azure.net`.
 
 ```bash tab="Example"
 --keyvault-name="MyKeyVault" 
-```
\ No newline at end of file
+```
+
+## Multifile Options
```
Comment (MadelineMurray) [inline docs/Reference/CLI/CLI-Syntax.md]: Does it make more sense to put this directly after the File Options section so the two methods for file-based signing are grouped together?

File: docs/Reference/CLI/CLI-Syntax.md [hunk 6]
```diff
@@ -282,4 +284,19 @@ Name of the vault to access. Sub-domain of `vault.azure.net`.
...
+
+### directory
+
+Path to the directory containing the keys and password files.
```
Comment (MadelineMurray) [inline docs/Reference/CLI/CLI-Syntax.md]: keys -> key 

Suggested change so it's parallel with password

File: docs/Tutorials/Multifile.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
+description: Signing transactions with multiple keys.
+<!--- END of page meta data -->
+
+# Using EthSigner with Multiple Key Signers
```
Comment (MadelineMurray) [inline docs/Tutorials/Multifile.md]: Key Signers -> Signing Keys? 

Suggestion only

File: docs/Tutorials/Multifile.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+## Start Besu 
+
+[Start Besu](https://besu.hyperledger.org/en/stable/HowTo/Get-Started/Starting-node/) with the 
+[`--rpc-http-port`](https://besu.hyperledger.org/en/stable/Reference/CLI/CLI-Syntax/#rpc-http-port)
+option set to `8590`. 
+
+!!! example
+    ```bash
+    besu --network=dev --miner-enabled --miner-coinbase=0xfe3b557e8fb62b89f4916b721be55ceb828dbd73 --rpc-http-cors-origins="all" --host-whitelist=* --rpc-http-enabled --rpc-http-port=8590 --data-path=/tmp/tmpDatdir
```
Comment (MadelineMurray) [inline docs/Tutorials/Multifile.md]: Not sure it makes sense to use `/tmp` path for a tutorial?

File: docs/Tutorials/Multifile.md [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    ```
+    
+## Create Passwords and Key Files
```
Comment (MadelineMurray) [inline docs/Tutorials/Multifile.md]: Passwords -> Password 

Same comment as above, needs to be parrallel with Key

File: docs/Tutorials/Multifile.md [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+option set to `8590`. 
+
+!!! example
+    ```bash
+    besu --network=dev --miner-enabled --miner-coinbase=0xfe3b557e8fb62b89f4916b721be55ceb828dbd73 --rpc-http-cors-origins="all" --host-whitelist=* --rpc-http-enabled --rpc-http-port=8590 --data-path=/tmp/tmpDatdir
+    ```
+    
+## Create Passwords and Key Files 
+
+Create a password file and V3 Keystore key for each node that needs to sign a transaction. The password files and V3 Keystore keys must follow the [naming convention](../Concepts/Multiple-Key-Files.md) and be available in the same directory.
```
Comment (MadelineMurray) [inline docs/Tutorials/Multifile.md]: node -> account?  

You can use multiple signing accounts with the same node.  From my understanding, that's why this change - so you don't have to run a seperate node for each signing account.

Comment (MadelineMurray) [reply docs/Tutorials/Multifile.md]: sign a transaction -> sign transactions

Comment (MadelineMurray) [reply docs/Tutorials/Multifile.md]: be available in -> be in 

I don't think available is required here

File: docs/Tutorials/Multifile.md [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+!!! attention "Password text file must not contain characters other than those used in your password"
```
Comment (MadelineMurray) [inline docs/Tutorials/Multifile.md]: attention -> caution 

I think this is the only place we use attention so I'd recommend changing to caution

File: docs/Tutorials/Multifile.md (hunk 10 above)
Comment (MadelineMurray) [reply docs/Tutorials/Multifile.md]: I think it probably makes sense to include the naming convention for the password file here - similar to what's done for the key file below.  eg, 

The password file must be named `[<prefix>]<accountAddress>.password`.

File: docs/Tutorials/Multifile.md [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+    
+Copy and paste the example JS script to a file (for example, `createKeyFile.js`) and replace the placeholders. 
+
+Use the JS script to display the text for the key file: 
+
+```bash
+node createKeyFile.js
+```
+
+Copy and paste the text to a file that is named `[<prefix>]<accountAddress>.key`. The file name must be identical to the password file.
```
Comment (MadelineMurray) [inline docs/Tutorials/Multifile.md]: ... to the password file. -> .... to the password file except for the `.key` suffix.  

Initially I thought both files had exactly the same name and was confused.

File: docs/Tutorials/Multifile.md [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+
+## Start EthSigner
+
+Start EthSigner with options specified as follows:
```
Comment (MadelineMurray) [inline docs/Tutorials/Multifile.md]: Start EthSigner with options specified as follows:  -> Start EthSigner with options: 

Suggested edit to remove redundant words

File: docs/Tutorials/Multifile.md [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+node createKeyFile.js
+```
+
+Copy and paste the text to a file that is named `[<prefix>]<accountAddress>.key`. The file name must be identical to the password file.
+
+## Start EthSigner
+
+Start EthSigner with options specified as follows: 
+
+* `chain-id` is the chain ID specified in the Besu genesis file.
```
Comment (MadelineMurray) [inline docs/Tutorials/Multifile.md]: Add a link to the Besu genesis file doc?  Suggestion only

File: docs/HowTo/Transactions/Make-Transactions.md (hunk 1 above)
Comment (MadelineMurray) [reply docs/HowTo/Transactions/Make-Transactions.md]: I agree with Lucas here - this is a related to the overall linking the signing key to the creating of key files.  ie, if we mention multifile here then we'd also be mentioning Azure and Hashicorp

File: docs/Reference/CLI/CLI-Syntax.md [hunk 15]
```diff
@@ -9,13 +9,15 @@ signs transaction with a key stored in an encrypted file or an external vault (f
 * `ethsigner [Options] file-based-signer [File Options]`
 * `ethsigner [Options] hashicorp-signer [Hashicorp Options]`
 * `ethsigner [Options] azure-signer [Azure Options]`
+* `ethsigner [Options] multifile-based-signer [Multifile Options]`
```
Comment (MadelineMurray) [inline docs/Reference/CLI/CLI-Syntax.md]: See comment below - probably makes sense to group the two file based options together

File: docs/Tutorials/Multifile.md (hunk 10 above)
Comment (MadelineMurray) [reply docs/Tutorials/Multifile.md]: And include an example - when following the tutorial steps I had to flick back through to the concept material to find an example to determine whether or not I should include the `0x` on the account address.

File: docs/HowTo/Transactions/Make-Transactions.md (hunk 1 above)
Comment (bgravenorst) [reply docs/HowTo/Transactions/Make-Transactions.md]: Done.

File: docs/HowTo/Transactions/Make-Transactions.md (hunk 2 above)
Comment (bgravenorst) [reply docs/HowTo/Transactions/Make-Transactions.md]: Done.

File: docs/Concepts/Multiple-Key-Files.md [hunk 16]
```diff
@@ -0,0 +1,23 @@
+description: Signing transactions with multiple keys.
+<!--- END of page meta data -->
+
+# Store Multiple Key Files
+
+EthSigner supports file-based signing using multiple V3 keystore files. Each account that signs transactions on the network requires a key file.
```
[the remark this answers is not in the document]
Comment (bgravenorst) [reply docs/Concepts/Multiple-Key-Files.md]: Done.

File: docs/Concepts/Multiple-Key-Files.md [hunk 17]
```diff
@@ -0,0 +1,23 @@
+description: Signing transactions with multiple keys.
+<!--- END of page meta data -->
+
+# Store Multiple Key Files
+
+EthSigner supports file-based signing using multiple V3 keystore files. Each account that signs transactions on the network requires a key file. 
+
+The key file must follow the appropriate naming convention, and have an accompanying password file. [Create the required files in a directory](../Tutorials/Multifile.md) and use the [multifile-based-signer](../Reference/CLI/CLI-Syntax.md#multifile-options) sub-command to specify the directory location.
```
[the remark this answers is not in the document]
Comment (bgravenorst) [reply docs/Concepts/Multiple-Key-Files.md]: Fixed.

File: docs/Tutorials/Multifile.md (hunk 10 above)
Comment (bgravenorst) [reply docs/Tutorials/Multifile.md]: Done.

File: docs/Tutorials/Multifile.md (hunk 12 above)
Comment (bgravenorst) [reply docs/Tutorials/Multifile.md]: Done.

File: docs/Tutorials/Multifile.md (hunk 13 above)
Comment (bgravenorst) [reply docs/Tutorials/Multifile.md]: Done.

File: docs/Tutorials/Multifile.md (hunk 14 above)
Comment (bgravenorst) [reply docs/Tutorials/Multifile.md]: Done.

File: docs/Reference/CLI/CLI-Syntax.md (hunk 15 above)
Comment (bgravenorst) [reply docs/Reference/CLI/CLI-Syntax.md]: Done.

File: docs/Concepts/Multiple-Key-Files.md (hunk 3 above)
Comment (bgravenorst) [reply docs/Concepts/Multiple-Key-Files.md]: Updated in the EthSigner doc. But I think we may need to raise a separate JI for making the Besu change.

File: docs/Reference/CLI/CLI-Syntax.md (hunk 5 above)
Comment (bgravenorst) [reply docs/Reference/CLI/CLI-Syntax.md]: Done.

File: docs/Tutorials/Multifile.md (hunk 11 above)
Comment (bgravenorst) [reply docs/Tutorials/Multifile.md]: Done.

File: docs/Reference/CLI/CLI-Syntax.md (hunk 6 above)
Comment (bgravenorst) [reply docs/Reference/CLI/CLI-Syntax.md]: Done.

File: docs/Tutorials/Multifile.md (hunk 7 above)
Comment (bgravenorst) [reply docs/Tutorials/Multifile.md]: Done.

File: docs/Tutorials/Multifile.md (hunk 8 above)
Comment (bgravenorst) [reply docs/Tutorials/Multifile.md]: Done. Fixed in both tutorials.

File: docs/Tutorials/Multifile.md (hunk 9 above)
Comment (bgravenorst) [reply docs/Tutorials/Multifile.md]: Done.

File: docs/Concepts/Multiple-Key-Files.md (hunk 3 above)
Comment (MadelineMurray) [reply docs/Concepts/Multiple-Key-Files.md]: Raised https://jira.hyperledger.org/browse/BESU-107 for the Besu change

File: mkdocs.yml [hunk 18]
```diff
@@ -65,11 +65,14 @@ nav:
             - Use Azure Key Vault: HowTo/Store-Keys/Use-Azure.md
         - Create External Signers: HowTo/Extending-EthSigner/Add-Signer.md
     - Tutorials:
-        - Start EthSigner: Tutorials/Start-EthSigner.md
+        - Start EthSigner with a Single Signer: Tutorials/Start-EthSigner.md
+        - Start EthSigner with Multiple Signers: Tutorials/Multifile.md
+    - Concepts:
+        - EthSigner Transaction Overview: Concepts/Overview.md
```
Comment (MadelineMurray) [inline mkdocs.yml]: EthSigner Transaction Overview -> Overview 

The name is pretty long in the left hand menu and not sure EthSigner Transaction adds any additional value

Comment (MadelineMurray) [reply mkdocs.yml]: It's much better with this content here than on the index page :-)

File: docs/Concepts/Multiple-Key-Files.md [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+
+In the example the account address is prefixed with the date. The `0x` portion of the account address must be removed.
+
+!!! example
+    `10-10-2019_78e6e236592597c09d5c137c2af40aecd42d12a2.key`
+    
+    `10-10-2019_78e6e236592597c09d5c137c2af40aecd42d12a2.password`
+    
+!!! tip
+    Use the [`export-address`](https://besu.hyperledger.org/en/latest/Reference/CLI/CLI-Subcommands/#export-address) Hyperledger Besu subcommand to obtain the public key address of the node.
```
Comment (MadelineMurray) [inline docs/Concepts/Multiple-Key-Files.md]: `public key address` -> `account address`

File: docs/Tutorials/Multifile.md [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+    echo -n "Type your password:";read -s password;echo -ne $password > [<prefix>]<accountAddress>.password;
+    ```
+    Enter the password when prompted.
+  
+Use the [web3.js library](https://github.com/ethereum/web3.js/) to create a key file where: 
+
+* `<AccountPrivateKey>` is the private key of the account with which EthSigner will sign transactions.  
+
+* `<Password>` is the password for the key file being created. The password must match the password saved in the  
+  password file created above.
```
Comment (MadelineMurray) [inline docs/Tutorials/Multifile.md]: There's a line break after _the_ in the published output - is there an extra space that needs removing here?

File: mkdocs.yml (hunk 18 above)
Comment (bgravenorst) [reply mkdocs.yml]: Done.

File: docs/Concepts/Multiple-Key-Files.md (hunk 19 above)
Comment (bgravenorst) [reply docs/Concepts/Multiple-Key-Files.md]: Done.

File: docs/Tutorials/Multifile.md (hunk 20 above)
Comment (bgravenorst) [reply docs/Tutorials/Multifile.md]: Done.
